HDDS-7812. Recon UI: Add visualization for container size distribution.

smitajoshi12 opened the pull request:
## What changes were proposed in this pull request?
Recon has an API endpoint to get container size distribution in Ozone. Add visualisation in Recon UI for this using histograms.
We will have to add another box that will display the percentages
Example
```
    {      
     "containerSize": 1024,       
      "count": 2     
    },
    {
      "containerSize": 16384,
      "count": 3
    },
    {
      "containerSize": 2097152,
      "count": 4
    },
    {
      "containerSize": 11270,
      "count": 5
    }
```
Total Count ➖ 2+3+4+5 = *14 Total Containers*
Percentage of Containers having upperbound 1024 = 2/14 *100 = 14.28%
Percentage of Containers having upperbound 16384 = 3/14 *100 = 21.42%
Percentage of Containers having upperbound 2097152 = 4/14 *100 = 28.5%
Percentage of Containers having upperbound 11270 = 5/14 *100 = 35.7%

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-7812

## How was this patch tested?
Manually
![image](https://user-images.githubusercontent.com/112169209/236488950-4be58c98-22a2-49d7-963b-998c7abb25db.png)


![image](https://user-images.githubusercontent.com/112169209/236740508-af1dd0ec-3893-421b-a378-bf469ff74ac3.png)

For Empty Response 

![image](https://user-images.githubusercontent.com/112169209/236740770-2ea922d3-495f-4603-a00b-3f9684967a1e.png)
------------------------------------
For Below Response with Actual Cluster Data
```
[
  {
    "containerSize": 536870912,
    "count": 1
  }
]
```

![image](https://github.com/apache/ozone/assets/112169209/61891434-8e30-46d4-bc46-75cf734f77ab)
---------------------------------------------------------------------------------------------------------------------------------------------

For Below Response with Actual Cluster Data
```
[
  {
    "containerSize": 536870912,
    "count": 1
  },
  {
    "containerSize": 1073741824,
    "count": 1
  }
]
```

![image](https://github.com/apache/ozone/assets/112169209/3f4ec764-7bdd-4036-bbc6-d644451d07e8)
----------------------------------------------------------------------------------------------------------------------------------------------
For Below Response with Actual Cluster Data
```
[
  {
    "containerSize": 536870912,
    "count": 1
  },
  {
    "containerSize": 1073741824,
    "count": 5
  }
]
```
![image](https://github.com/apache/ozone/assets/112169209/f9b73080-3863-49a9-a2dc-55cf019dc8a2)
------------------------------------------------------------------------------------------------------------------------------------------------------
For Below Response with Actual Cluster Data
```
[
  {
    "containerSize": 536870912,
    "count": 1
  },
  {
    "containerSize": 1073741824,
    "count": 10
  }
]
```
![image](https://github.com/apache/ozone/assets/112169209/461402d3-bf07-49de-a23c-883283aee9a3)
---------------------------------------------------------------------------------------------------------------------------------------------------
Sample Data for Testing
```
[
  {
    "containerSize": 536870912,
    "count": 1
  },
  {
    "containerSize": 1073741824,
    "count": 19
  },
  {
    "containerSize": 5306870912,
    "count": 5
  },
  {
    "containerSize": 10073741824,
    "count": 20 
  },
  {
    "containerSize": 100073741824,
    "count": 14
  }
]
```
Color 
Blue :- Count 20
Orange  :- Count 19
Green :- Count 14
Red :- Count 5
Violet :- Count 1

![image](https://github.com/apache/ozone/assets/112169209/1cd18845-c64a-4475-a102-e211781f1b3a)
---------------------------------------------------------------------------------
Sample Data for Testing
```
[
  {
    "containerSize": 5306870912,
    "count": 5
  },
  {
    "containerSize": 10073741824,
    "count": 20
  },
  {
    "containerSize": 1000073741824,
    "count": 14
  }
]
```
Color 
Blue :- Count 20
Green  :- Count  5
Orange :- Count 14

![image](https://github.com/apache/ozone/assets/112169209/c69c91c0-9aaf-45b2-8803-935ae4b2559b)
---------------------------------------------------------

ArafatKhan2198 commented:
@smitajoshi12 Can you please add the screenshot of the latest Pie Chart that we had discussed !!

smitajoshi12 commented:
@ArafatKhan2198  These are latest changes of patch.

ArafatKhan2198 commented:
@smitajoshi12 These screenshots indicate that the size of some containers is less than 512 MB, which is not possible according to the backend. To address this, I implemented changes to the backend that group all containers with sizes smaller than 512 MB into the range of 0-512 MB. Please pull in the latest changes and populate a local ozone cluster with data and take scereenshots of those.

smitajoshi12 commented:
> @smitajoshi12 These screenshots indicate that the size of some containers is less than 512 MB, which is not possible according to the backend. To address this, I implemented changes to the backend that group all containers with sizes smaller than 512 MB into the range of 0-512 MB. Please pull in the latest changes and populate a local ozone cluster with data and take scereenshots of those.

@ArafatKhan2198 
Added new screenshots.

dombizita commented on hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/insights.tsx:
```diff
@@ -36,10 +37,17 @@ interface IFileCountResponse {
   isLoading: boolean;
   fileCountsResponse: IFileCountResponse[];
+  containerCountResponse: IContainerCountResponse[];
   plotData: Plotly.Data[];
+  containerData: Plotly.Data[];
```
as we are having both container and file count size task now, we should rename the previous variable, so it is easier to understand it. 
```suggestion
  fileCountData: Plotly.Data[];
  containerCountData: Plotly.Data[];
```

smitajoshi12 replied:
@dombizita 
Completed Change Suggested by you.

dombizita commented on hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/insights.tsx:
```diff
@@ -155,12 +165,45 @@ export class Insights extends React.Component<Record<string, object>, IInsightsS
         return `${lowerbound} - ${upperbound}`;
       });
+      
+      const xyConatainerCountMap: Map<number, number> = containerCountResponse.reduce(
```
```suggestion
      const xyContainerCountMap: Map<number, number> = containerCountResponse.reduce(
```

dombizita commented on hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/insights.tsx:
```diff
@@ -155,12 +165,45 @@ export class Insights extends React.Component<Record<string, object>, IInsightsS
+      
+      const xyConatainerCountMap: Map<number, number> = containerCountResponse.reduce(
+        (map: Map<number, number>, current) => {
+          const fileSize = current.containerSize;
```
```suggestion
          const containerSize = current.containerSize;
```

smitajoshi12 replied:
@dombizita 
Completed Change Suggested by you.

dombizita commented on hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/insights.tsx:
```diff
@@ -155,12 +165,45 @@ export class Insights extends React.Component<Record<string, object>, IInsightsS
         plotData: [{
           type: 'bar',
           x: xValues,
           y: Array.from(xyMap.values()),
```
```suggestion
          y: Array.from(xyFileCountMap.values()),
```

smitajoshi12 replied:
@dombizita 
Completed Change Suggested by you.

dombizita commented on hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/insights.tsx:
```diff
@@ -155,12 +165,45 @@ export class Insights extends React.Component<Record<string, object>, IInsightsS
       this.setState({
         plotData: [{
           type: 'bar',
           x: xValues,
```
```suggestion
          x: xFileCountValues,
```

dombizita commented on hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/insights.tsx:
```diff
@@ -215,69 +264,107 @@ export class Insights extends React.Component<Record<string, object>, IInsightsS
+                          </Col>
+                        </Row>
+                      </div> :
+                      <div>No data to visualize Container size distribution. Add files to Ozone to see a visualization on Container size distribution.</div>)}
```
```suggestion
                      <div>No data to visualize container size distribution. Add files to Ozone to see a visualization on container size distribution.</div>)}
```

smitajoshi12 replied:
@dombizita 
Completed Change Suggested by you.

dombizita reviewed (commented):
thanks for working on this @smitajoshi12, I added a few comments inline regarding naming, most of them are because now we have both file and container size distribution tasks, so the variables should be named inline with what are they used for, please take a look at my suggestions

dombizita commented:
can you also review this patch @ArafatKhan2198?

ArafatKhan2198 commented:
Can you please make the allingment proper, the some of the letters get hidden :-
<img width="175" alt="image" src="https://github.com/apache/ozone/assets/98023601/5a8958a7-a6f0-46ce-b212-00ec03aa7f0a">

ArafatKhan2198 commented:
Please ensure that the text color is white for those specific colors where it currently appears as black.

<img width="183" alt="image" src="https://github.com/apache/ozone/assets/98023601/a22f7366-9770-4431-9808-f895c27c961c">

ArafatKhan2198 commented:
Change the message to ➖ ”No data available for container size distribution visualization. Add files to Ozone.”
which sounds more simple and short

<img width="1651" alt="image" src="https://github.com/apache/ozone/assets/98023601/72ce083d-ff71-40f0-9b5f-b280ec7fe8d4">

smitajoshi12 commented:
> Change the message to heavy_minus_sign ”No data available for container size distribution visualization. Add files to Ozone.” which sounds more simple and short
> 
> <img alt="image" width="1651" src="https://user-images.githubusercontent.com/98023601/240267055-72ce083d-ff71-40f0-9b5f-b280ec7fe8d4.png">



> Can you please make the alinement proper, some of the letters get hidden :- <img alt="image" width="175" src="https://user-images.githubusercontent.com/98023601/240265824-5a8958a7-a6f0-46ce-b212-00ec03aa7f0a.png">

Completed Changes Suggested by Arafat.

smitajoshi12 commented:
> Change the message to heavy_minus_sign ”No data available for container size distribution visualization. Add files to Ozone.” which sounds more simple and short
> 
> <img alt="image" width="1651" src="https://user-images.githubusercontent.com/98023601/240267055-72ce083d-ff71-40f0-9b5f-b280ec7fe8d4.png">

Completed Changes Suggested by Arafat.

![image](https://github.com/apache/ozone/assets/112169209/2c495c3c-336e-4a0e-ab44-3630a247037d)

smitajoshi12 commented:
> Please ensure that the text color is white for those specific colors where it currently appears as black.
> 
> <img alt="image" width="183" src="https://user-images.githubusercontent.com/98023601/240266391-a22f7366-9770-4431-9808-f895c27c961c.png">

@ArafatKhan2198 
In Disk Usage it is internally manged by library we could not modify color of tool-tip as per background color it will vary.

smengcl commented:
Thanks @smitajoshi12 for the patch. Thanks @dombizita for reviewing this.
